src/blocks/file-blocks/code/index.tsx

```diff
@@ -1,39 +1,188 @@
+    load: () =>
+      import("@codemirror/lang-javascript").then(
+        (m) => new LanguageSupport(m.javascriptLanguage)
+      ),
+  },
+  {
+    name: "JavaScript + JSX",
+    extensions: ["jsx"],
+    load: () =>
+      import("@codemirror/lang-javascript").then(
+        (m) => new LanguageSupport(m.jsxLanguage)
+      ),
+  },
+  {
+    name: "TypeScript",
+    extensions: ["ts"],
+    load: () =>
+      import("@codemirror/lang-javascript").then(
+        (m) => new LanguageSupport(m.typescriptLanguage)
+      ),
+  },
+  {
+    name: "TypeScript + TSX",
+    extensions: ["tsx"],
+    load: () =>
+      import("@codemirror/lang-javascript").then(
+        (m) => new LanguageSupport(m.tsxLanguage)
+      ),
+  },
+].map(LanguageDescription.of);
+
+const languageConf = new Compartment();
+
+let doc: string;
+
+const extensions = [
+  lineNumbers(),
+  highlightActiveLineGutter(),
+  highlightSpecialChars(),
+  history(),
+  foldGutter(),
+  drawSelection(),
+  dropCursor(),
+  EditorState.allowMultipleSelections.of(true),
+  indentOnInput(),
+  defaultHighlightStyle.fallback,
+  bracketMatching(),
+  closeBrackets(),
+  autocompletion(),
+  rectangularSelection(),
+  highlightActiveLine(),
+  highlightSelectionMatches(),
+  keymap.of([
+    ...closeBracketsKeymap,
+    ...defaultKeymap,
+    ...searchKeymap,
+    ...historyKeymap,
+    ...foldKeymap,
+    ...commentKeymap,
+    ...completionKeymap,
+    ...lintKeymap,
+  ]),
+
+  languageConf.of([]),
+
+  EditorView.updateListener.of((v) => {
+    if (!v.docChanged) return;
+    doc = v.state.doc.sliceString(0);
+  }),
+];
 
 export default function (props: FileBlockProps) {
   const {
     content,
     context: { path },
+    onRequestUpdateContent,
   } = props;
 
-  const extension = path.includes(".") ? path.split(".").pop() : undefined;
-  const language = languageOfExtension(extension);
+  doc = content;
+  const editorRef = React.useRef<HTMLDivElement>(null);
+
+  React.useEffect(() => {
+    if (!editorRef.current) return;
+    const state = EditorState.create({
+      doc: content,
+      extensions,
+    });
+    const view = new EditorView({
+      state,
+      parent: editorRef.current,
+    });
+
+    const language = LanguageDescription.matchFilename(
+      languageDescriptions,
+      path
+    );
+
+    if (language) {
+      language.load().then((lang) => {
+        view.dispatch({
+          effects: languageConf.reconfigure(lang),
+        });
+      });
+    }
+  }, [editorRef.current]);
 
   return (
-    <div className="code">
-      <SyntaxHighlighter
-        language={language}
-        useInlineStyles={false}
-        showLineNumbers
-        lineNumberStyle={{ opacity: 0.45 }}
+    <div className="position-relative height-full">
+      <div
+        className="width-full"
+        key={content}
+        style={{ height: "100vh" }}
+        ref={editorRef}
+      />
+      <button
+        className="btn btn-primary position-absolute right-4 top-4 z-10"
```

Comment:
Personal preference, but I think this should be `fixed` so that it stays with you as you scroll down the potentially long file.

Reply:
makes sense! (I copied this from the Excalidraw block and didn't try it on a long file.)

Reply:
great call! chiming in that you might need to make it absolutely positioned in a full-height block that scrolls, otherwise `fixed` will be relative to the entire prototype (for example blocks)

Reply:
or you could create a new stacking context (which we should probably just do automatically for the example block wrapper)

Reply:
done!